src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,344 @@
+        break;
+
+      default:
+        break;
+    }
+
+  };
+
+  public displayLinkOptions = () => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (autoCompleteOptions) {
+      const suggestions: string[] = [];
+      const parameters = autoCompleteOptions.parameters;
+      if (parameters) {
+        const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+        if (parametersWithVerb) {
+          parametersWithVerb.links.forEach((value: string) => {
+            suggestions.push(value);
+          });
+
+          this.setState({
+            filteredSuggestions: suggestions,
+            suggestions,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  public getQueryParameters = () => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (autoCompleteOptions) {
+      const suggestions: string[] = [];
+      const parameters = autoCompleteOptions.parameters;
+      if (parameters) {
+        const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+        if (parametersWithVerb) {
+          parametersWithVerb.values.forEach((value: any) => {
+            suggestions.push(value.name);
+          });
+
+          this.setState({
+            filteredSuggestions: suggestions,
+            suggestions,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  public componentDidUpdate = (prevProps: IAutoCompleteProps) => {
+    if (prevProps.autoCompleteOptions !== this.props.autoCompleteOptions) {
+      if (this.props.autoCompleteOptions) {
+        this.performLocalSearch(this.props.sampleQuery.sampleUrl);
+      }
+    }
+  }
+
+  private getParameterEnums = (url: string) => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (!autoCompleteOptions) {
+      return;
+    }
+    const parameters = autoCompleteOptions.parameters;
+    const param = url.split('$').pop()!.split('=')[0];
+    if (parameters) {
+      const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+      if (parametersWithVerb) {
+        const section = parametersWithVerb.values.find((k: { name: string; }) => {
+          return k.name === `$${param}`;
+        });
+
+        let list: string[] = [];
+        if (section && section.items && section.items.length > 0) {
+          list = section.items;
+          this.setState({
+            filteredSuggestions: list,
+            suggestions: list,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  private filterSuggestions(userInput: string, previousUserInput: string, compare: string, suggestions: string[]) {
+    let compareString = userInput.replace(previousUserInput, '');
```

Comment:
If `previousUserInput` is `a` and `userInput` is `aa`, `compareString` will be empty string as `replace` function replaces all instances of the first parameter. Is this an edge case that needs to be covered?

Reply:
Yes. Interesting that it actually comes up when you have a query with multiple query parameters with similar values. e.g
`https://graph.microsoft.com/v1.0/me/messages?$select=id,subject&orderby=su` fixed it with a slice function instead of the loop I was running.